bpo-10544: Deprecate "yield" in comprehensions and generator expressions.

serhiy-storchaka opened the pull request:
<!-- issue-number: bpo-10544 -->
https://bugs.python.org/issue10544
<!-- /issue-number -->

serhiy-storchaka committed:
bpo-10544: Disallow "yield" in comprehensions and generator expressions.

serhiy-storchaka committed:
Merge branch 'master' into deprecate-yield-in-comprehensions

serhiy-storchaka committed:
Emit a warning istead of an error.

ncoghlan commented:
@serhiy-storchaka Do you mind if I just edit some of the docs wording directly? I think that will be easier than trying to describe the edits I would recommend.

serhiy-storchaka commented:
I will be grateful if you do this.

ncoghlan committed:
Reword language reference updates

- updated the comprehension scoping definition to use
  some of the words from the generator expression definition
- added appropriate versionadded/changed directives
-

ncoghlan committed:
Fix typo in markup changes

ncoghlan committed:
Reword What's New notice

ncoghlan committed:
Reword NEWS entry

serhiy-storchaka commented on Doc/reference/expressions.rst:
```diff
@@ -201,6 +210,12 @@ or :keyword:`await` expressions it is called an
 See also :pep:`530`.
 
+.. versionadded:: 3.6
+   Asynchronous comprehensions were introduced
```
Missed a period.

ncoghlan committed:
Describe specific consequences in What's New entry

serhiy-storchaka commented on Doc/reference/expressions.rst:
```diff
@@ -183,11 +183,20 @@ by considering each of the :keyword:`for` or :keyword:`if` clauses a block,
+and then passed as an argument to the implictly nested scope. Subsequent
+:keyword:`for` clauses cannot be evaluated in the enclosing scope since they
+may depend on the previous :keyword:`for` loop. For example:
+``[x*y for x in range(10) for y in bar(x)]``.
```
Maybe use more concrete real-world example? E.g. `for y in range(x+1, 10)`.

ncoghlan replied:
I lifted this one directly from the generator expression docs, but changing it in both places would definitely be reasonable.

serhiy-storchaka commented on Doc/reference/expressions.rst:
```diff
@@ -201,6 +210,12 @@ or :keyword:`await` expressions it is called an
+.. versionadded:: 3.6
+   Asynchronous comprehensions were introduced
+
+.. versionchanged:: 3.7
```
Shouldn't be `deprecated` or `deprecated-removed`?

ncoghlan replied:
Aye, it should - I just forgot there was a specific directive for deprecations. I'll go through and fix all the cases.

ncoghlan commented:
OK, I've made my documentation edits.

@gvanrossum I think it would be helpful if you could review the proposed language reference updates directly, or else nominate someone else to do so.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -563,7 +563,10 @@ Deprecated
   expressions won't attempt to interleave their implicit output with the output
   from any explicit yield expressions.
-  (Contributed by Serhiy Storchaka in :issue:`10544`.)
+  
```
Trailing spaces.

ncoghlan committed:
Remove trailing whitespace

ncoghlan committed:
Incorporate Serhiy's review comments

serhiy-storchaka commented on Doc/reference/expressions.rst:
```diff
@@ -211,9 +211,9 @@ suspend the execution of the coroutine function in which it appears.
 
-.. versionchanged:: 3.7
+.. deprecated:: 3.7
    ``yield`` and ``yield from`` deprecated in the implicitly nested scope
```
Missed a period.

serhiy-storchaka committed:
Add an explicit PyErr_Clear() and improve the comment.

serhiy-storchaka commented on Doc/whatsnew/3.7.rst:
```diff
@@ -556,6 +556,18 @@ Other CPython Implementation Changes
+  in comprehensions and generator expressions (aside from the outermost iterable
+  defined in the leftmost :keyword`for` clause). This ensures that comprehensions
+  always immediately return a container of the appropriate type (rather than
+  potentially returning a :term:`generator-iterator` object), while generator
```
> Warning, treated as error:
> /home/travis/build/python/cpython/Doc/whatsnew/3.7.rst:559:term not in glossary: generator-iterator

serhiy-storchaka commented on Doc/reference/expressions.rst:
```diff
@@ -198,6 +210,12 @@ or :keyword:`await` expressions it is called an
+   Asynchronous comprehensions were introduced.
+
+.. deprecated:: 3.7
+   ``yield`` and ``yield from`` deprecated in the implicitly nested scope
```
Missed a period.

serhiy-storchaka committed:
Fix minor doc errors.

gvanrossum commented on Doc/reference/expressions.rst:
```diff
@@ -183,8 +183,20 @@ by considering each of the :keyword:`for` or :keyword:`if` clauses a block,
 
-Note that the comprehension is executed in a separate scope, so names assigned
-to in the target list don't "leak" into the enclosing scope.
+However, aside from the leftmost :keyword:`for` clause, the comprehension is
```
I think "leftmost `for` clause" is still ambiguous. Does this include the <expr> in `for a[<expr>] in ...`? It should not, but I would assume it's part of "the `for` clause." Maybe say "the iterable in the leftmost `for` clause"?

ncoghlan replied:
Using just "iterable" ended up looking odd in context (since executing an iterable could be taken to mean iterating over it).

However, "the iterable expression in the leftmost :keyword:`for` clause" reads reasonably well to me, and I also made it explicit that the leftmost filter condition is evaluated in the nested scope.

It may be worth our while to explicitly define "outermost iterable" as a formal term in the glossary, but I think that would be better handled as a distinct docs RFE.

gvanrossum commented on Doc/whatsnew/3.7.rst:
```diff
@@ -556,6 +556,18 @@ Other CPython Implementation Changes
 
+* Yield expressions (both ``yield`` and ``yield from`` clauses) are now deprecated
+  in comprehensions and generator expressions (aside from the outermost iterable
+  defined in the leftmost :keyword:`for` clause). This ensures that comprehensions
```
I'd drop "defined" here.

ncoghlan replied:
Based on the changes I made to the language reference, I replaced "the outermost iterable defined in" with "the iterable expression in" here.

gvanrossum reviewed (commented):
Here are some suggestions to tighten the docs. Thanks to both of you for taking care of this!

ncoghlan committed:
Clarify eager evaluation of leftmost iterable

- consistently use "leftmost" (never "outermost")
- explicitly exclude the filter condition from eager evaluation
- better align wording for comprehensions and generator expressions

ncoghlan committed:
Tweak What's New wording

gvanrossum commented:
Docs LGTM now.

ncoghlan commented on Python/symtable.c:
```diff
@@ -1754,6 +1753,36 @@ symtable_handle_comprehension(struct symtable *st, expr_ty e,
+                   to get a more accurate error report */
+                PyErr_Clear();
+                PyErr_SetObject(PyExc_SyntaxError, msg);
+                PyErr_SyntaxLocationObject(st->st_filename,
```
Nice!

serhiy-storchaka commented:
Ping.

serhiy-storchaka commented:
Thank you for your documentation changes @ncoghlan!
